<a href="https://colab.research.google.com/github/kennis000/BTech_Ken/blob/main/M5_Project_Exploratory_Data_Analysis_on_Wine_Reviews_KW.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## **Module 5 Project: Exploratory Data Analysis on Wine Reviews**
By: Ken Wilson

Wine tasting is the art of evaluating wines through sight, smell, and taste. It involves assessing color, identifying aromas, and analyzing flavors, acidity, and texture. Subjective yet informative, it helps judge quality, age potential, and personal preferences, making it a nuanced exploration of wine's intricate characteristics.

### **Project Overview**

As part of an ongoing desire to determine which characteristics make a wine great and which ones do not, the stake holders would like the following questions explored and answered:

    1. Which country produces wine with the most points, on average? (descriptive statistics)
    2. Which taster gives the lowest scores (points), on average?
    3. Which variety of wine is the most expensive, on average?
    4. Which year of wines has the best score (points), on average?
    5. Do reviews with the word "depth" in them tend to get better than average or worse than average points?
    6. Do reviews with the word "fruity" in them tend to get better than average or worse than average points?
    7. Do reviews with the word "herbal" in them tend to get better than average or worse than average points?
    8. Do reviews with more letters award more or less points, on average?
    9. Which region of the province Sicily & Sardinia produces the best wine, on average?

### About the Dataset:

The dataset used in this project comprises wine reviews scraped from the Wine Enthusiast website (a magazine and ecommerce business) by Kaggle user zackthoutt during the week of June 15th, 2017. It includes various attributes such as the wine's country of origin, variety, points awarded, price, and detailed descriptions, among others.

Download the data set [wine.csv](https://btech.instructure.com/courses/610804/files/118105297/download?download_frd=1).

### This is a peek at the data we will be working with.

In [ ]:
import pandas as pd
df = pd.read_csv('/content/wine.csv')
display(df.info())
display(df.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 129971 entries, 0 to 129970
Data columns (total 11 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   country                129908 non-null  object 
 1   description            129971 non-null  object 
 2   designation            92506 non-null   object 
 3   points                 129971 non-null  int64  
 4   price                  120975 non-null  float64
 5   province               129908 non-null  object 
 6   taster_name            103727 non-null  object 
 7   taster_twitter_handle  98758 non-null   object 
 8   title                  129971 non-null  object 
 9   variety                129970 non-null  object 
 10  winery                 129971 non-null  object 
dtypes: float64(1), int64(1), object(9)
memory usage: 10.9+ MB


None

country  \
0     Italy   
1  Portugal   
2        US   
3        US   
4        US   

                                                                                                                                                                                                                                                 description  \
0                                                                               Aromas include tropical fruit, broom, brimstone and dried herb. The palate isn't overly expressive, offering unripened apple, citrus and dried sage alongside brisk acidity.   
1                        This is ripe and fruity, a wine that is smooth while still structured. Firm tannins are filled out with juicy red berry fruits and freshened with acidity. It's  already drinkable, although it will certainly be better from 2016.   
2                                                                 Tart and snappy, the flavors of lime flesh and rind dominate. Some green pineapple pokes through, with crisp acidity underscoring the flavors. The wine was all stainless-steel fermented.   
3                                                    Pineapple rind, lemon pith and orange blossom start off the aromas. The palate is a bit more opulent, with notes of honey-drizzled guava and mango giving way to a slightly astringent, semidry finish.   
4  Much like the regular bottling from 2012, this comes across as rather rough and tannic, with rustic, earthy, herbal characteristics. Nonetheless, if you think of it as a pleasantly unfussy country wine, it's a good companion to a hearty winter stew.   

                          designation  points  price           province  \
0                        Vulkà Bianco      87    NaN  Sicily & Sardinia   
1                            Avidagos      87   15.0              Douro   
2                                 NaN      87   14.0             Oregon   
3                Reserve Late Harvest      87   13.0           Michigan   
4  Vintner's Reserve Wild Child Block      87   65.0             Oregon   

          taster_name taster_twitter_handle  \
0       Kerin O’Keefe          @kerinokeefe   
1          Roger Voss            @vossroger   
2        Paul Gregutt           @paulgwine    
3  Alexander Peartree                   NaN   
4        Paul Gregutt           @paulgwine    

                                                                                 title  \
0                                                    Nicosia 2013 Vulkà Bianco  (Etna)   
1                                        Quinta dos Avidagos 2011 Avidagos Red (Douro)   
2                                        Rainstorm 2013 Pinot Gris (Willamette Valley)   
3                  St. Julian 2013 Reserve Late Harvest Riesling (Lake Michigan Shore)   
4  Sweet Cheeks 2012 Vintner's Reserve Wild Child Block Pinot Noir (Willamette Valley)   

          variety               winery  
0     White Blend              Nicosia  
1  Portuguese Red  Quinta dos Avidagos  
2      Pinot Gris            Rainstorm  
3        Riesling           St. Julian  
4      Pinot Noir         Sweet Cheeks

### Quick stats on points for included data:
* 129,971 wine reviews are included in this data.
* The average number of points given to all wines is 88.45.
* The median is 88.00.
* The maximum is 100.
* The minimum is 80.

In [ ]:
df['points'].describe()

count    129971.000000
mean         88.447138
std           3.039730
min          80.000000
25%          86.000000
50%          88.000000
75%          91.000000
max         100.000000
Name: points, dtype: float64

### **1. Highest Scoring Average by Country**

Wines from 43 different countries were included in this set of reviews.
### <u>England</u> is the country that scores the most points, on average.
* 91.58 is the average score for wines from England
* 74 wines from England were reviewed.

India and Austria follow close behind.

In [ ]:
# Number of countries having wines reviewed
print('\nNumber countries having wines reviewed: ' + str(df['country'].nunique()))

### Top 10 countries by average points
print('\n\nTop 10 countries by average points:\n')
df[['country','points']].groupby('country').agg(avg_points = ('points', 'mean'), median_points = ('points', 'median'), reviews_count = ('country', 'count')).sort_values(by = 'avg_points', ascending = False).reset_index().head(10)


Number countries having wines reviewed: 43


Top 10 countries by average points:



country  avg_points  median_points  reviews_count
0     England   91.581081           91.0             74
1       India   90.222222           90.0              9
2     Austria   90.101345           90.0           3345
3     Germany   89.851732           90.0           2165
4      Canada   89.369650           90.0            257
5     Hungary   89.191781           89.0            146
6       China   89.000000           89.0              1
7      France   88.845109           89.0          22093
8  Luxembourg   88.666667           88.5              6
9   Australia   88.580507           89.0           2329

### **2. The taster giving the lowest average amount of points**

19 individual tasters have wine reviews included in this data.

### <u>Alexander Peartree</u> is the taster who gave the lowest scores, on average.
* He gave reviews on 415 of these wines.
* The average score given by A. Peartree is:  85.86
##### Carrie Dykes's and Susan Kostrzewa's awarded points that average relatively low, as well.

In [ ]:
### Number of tasters that participated:
print('\nNumber of tasters that participated: ' + str(df['taster_name'].nunique()) + '\n')

### Lowest 5 tasters' average given points
print("\nLowest 5 tasters' average given points:\n")
display(df[['taster_name','points']].groupby('taster_name').agg(avg_points = ('points', 'mean'), median_points = ('points', 'median'),reviewed_count = ('taster_name', 'count')).sort_values(by = 'avg_points').reset_index().head())


Number of tasters that participated: 19


Lowest 5 tasters' average given points:



taster_name  avg_points  median_points  reviewed_count
0  Alexander Peartree   85.855422           86.0             415
1        Carrie Dykes   86.395683           87.0             139
2     Susan Kostrzewa   86.609217           87.0            1085
3         Fiona Adams   86.888889           87.0              27
4   Michael Schachner   86.907493           87.0           15134

### Bonus data:
Roger Voss contributed the most reviews by far!
* 25,514 of these reviews are by him! 19.6% of 129,971.
* R. Voss gave an average score of 88.71 (2.85 points higher than A. Peartree's average given score).

In [ ]:
print('\nTop 5 Tasters with the most reviews:\n')
display(df[['taster_name', 'points']].groupby('taster_name').agg(reviewed_count = ('taster_name','count'), avg_points = ('points', 'mean')).sort_values(by = 'reviewed_count', ascending = False).reset_index().head())


Top 5 Tasters with the most reviews:



taster_name  reviewed_count  avg_points
0         Roger Voss           25514   88.708003
1  Michael Schachner           15134   86.907493
2      Kerin O’Keefe           10776   88.867947
3     Virginie Boone            9537   89.213379
4       Paul Gregutt            9532   89.082564

### **3. Most expensive wine variety**

### <u>Ramisco</u> is the most expensive variety of wine reveiwed in this data, on average, at \$495.
A caveat to note is that <u>the 3 most expensive varieties on average (Ramisco, Terrantez, & Francisa) had only a single review in this set</u>. So, these single-bottle prices at \$495, \$236, & \$160 respectively may not represent the prices of these varieties very well.

***Muscadelle** at an average price of $111.20 is the most expensive variety having 10 or more reviews.



In [ ]:
print('\nMost expensive varieties, on average.\n')
display(df[['variety','price','points']].groupby('variety').agg(avg_price = ('price', 'mean'), variety_count = ('variety', 'count'), avg_points = ('points', 'mean')).sort_values(by = 'avg_price', ascending = False).reset_index().head(10))


Most expensive varieties, on average.



variety   avg_price  variety_count  avg_points
0           Ramisco  495.000000              1   93.000000
1         Terrantez  236.000000              1   95.000000
2          Francisa  160.000000              1   91.000000
3  Rosenmuskateller  150.000000              2   92.500000
4   Malbec-Cabernet  113.333333              3   91.666667
5  Tinta Negra Mole  112.000000              1   92.000000
6        Muscadelle  111.200000             10   92.500000
7              Bual  100.000000              7   94.142857
8             Debit   98.333333              3   89.000000
9   Cabernet-Shiraz   83.500000              2   91.000000

### Bonus data:
* The most expensive bottle, at \$3,300, was a 'Bordeaux-style Red Blend' from France titled *'Château les Ormes Sorbet 2013 Médoc'*.
* The highest priced wine did not receive the highest score.
* The highest priced wine scored 88 points where the second most expensive wine, a Pinot Noir at \$2,500, earned 96 points.

In [ ]:
### Most expensive wine tasted
print('\nThe 5 most expensive wines.\n')
display(df.drop(columns = ['description', 'designation', 'taster_twitter_handle']).sort_values(by = 'price', ascending = False).head())


The 5 most expensive wines.



country  points   price    province    taster_name  \
80290   France      88  3300.0    Bordeaux     Roger Voss   
98380   France      96  2500.0    Burgundy     Roger Voss   
15840   France      96  2500.0    Bordeaux     Roger Voss   
120391      US      91  2013.0  California  Matt Kettmann   
65352   France      97  2000.0    Bordeaux     Roger Voss   

                                                          title  \
80290                      Château les Ormes Sorbet 2013  Médoc   
98380            Domaine du Comte Liger-Belair 2010  La Romanée   
15840                              Château Pétrus 2014  Pomerol   
120391  Blair 2013 Roger Rose Vineyard Chardonnay (Arroyo Seco)   
65352                              Château Pétrus 2011  Pomerol   

                         variety                         winery  
80290   Bordeaux-style Red Blend       Château les Ormes Sorbet  
98380                 Pinot Noir  Domaine du Comte Liger-Belair  
15840   Bordeaux-style Red Blend                 Château Pétrus  
120391                Chardonnay                          Blair  
65352   Bordeaux-style Red Blend                 Château Pétrus

### **4. Year with the highest average points given**

### <u>1969</u> was the vintage year that received the highest average points (98.00) for wine in this data.
A caveat to note is that <u>wines of vintage 1969 had only a single review in this set</u>. Based on this point, we cannot say that wines from 1969 tended to earn higher scores.

Over 81 percent of the reviews in this data are from vintage years 2005 - 2016. These are the years where the average points may more appropriately represent the quality of wine from the given year.

**2013** is the year in this range (2005 - 2016) with the highest average score (88.95), having 15,853 reviews for that year's wines.

In [ ]:
### Extract vintage year from title and store in column 'year'
df['year'] = df['title'].str.extract(r'(\d{4})')[0]

### Copy average points and count by year into new dataframe 'points_by_year'
points_by_year = df[['year', 'points']].groupby('year').agg(avg_points = ('points', 'mean'), reviewed_count = ('year', 'count'))

In [ ]:
### Highest average score by year
print('\nThe 10 highest scoring years by average points:\n')
display(points_by_year.sort_values(by = 'avg_points', ascending = False).reset_index().head(10))


The 10 highest scoring years by average points:



year  avg_points  reviewed_count
0  1969        98.0               1
1  1973        96.0               1
2  1952        95.5               2
3  1961        95.0               1
4  1945        95.0               1
5  1966        95.0               2
6  1957        95.0               1
7  1927        95.0               2
8  1968        95.0               1
9  1982        95.0               1

In [ ]:
### The 20 years having the most number of wines reviewed
print('\n\nThe 20 years having the most number of wines reviewed:\n')
display(points_by_year.drop(columns = 'avg_points').sort_values(by = 'reviewed_count', ascending = False).head(20).reset_index())



The 20 years having the most number of wines reviewed:



year  reviewed_count
0   2013           15853
1   2012           15723
2   2014           15568
3   2011           12531
4   2010           12149
5   2015           10041
6   2009            9864
7   2008            7426
8   2007            7041
9   2006            5772
10  2016            3690
11  2005            3610
12  2004            1772
13  2000             758
14  2001             691
15  1999             641
16  2003             597
17  1998             550
18  2002             365
19  1997             303

In [ ]:
### The 10 highest scoring years by average points having at least 3600 reviews
print('\n\nThe 10 highest scoring years by average points having at least 3600 reviews:\n')
display(points_by_year[points_by_year['reviewed_count'] >= 3600].sort_values('avg_points', ascending = False).head(10).reset_index())



The 10 highest scoring years by average points having at least 3600 reviews:



year  avg_points  reviewed_count
0  2013   88.951870           15853
1  2014   88.866971           15568
2  2012   88.773771           15723
3  2015   88.519869           10041
4  2005   88.353186            3610
5  2009   88.335766            9864
6  2011   88.278429           12531
7  2010   88.264137           12149
8  2006   88.216216            5772
9  2008   88.205629            7426

### **5. Wines described using the word "depth".**
### Wines described using the word "<u>depth</u>" get <u>MORE</u> points on average.
* Reviews with the word "depth" averaged **90.11** points.
* All wines together averaged 88.45 points.
* Being described with "depth" resulted in an average 1.66 point advantage.
* The number of wines that have the word "depth" in their descriptions is 2560.

In [ ]:
### If the description contains the word "depth", mark True in a new column named 'described_depth'
df['described_depth'] = df['description'].str.contains(r'[Dd]epth')

### Count the number of rows with "depth"
print('Number of wines with "depth": ' + str(df[df['described_depth'] == True].shape[0]))

### Average the points given to those with "depth".
avg_depth = round(df[df['described_depth'] == True]['points'].mean(),2)
print('Average points having "depth": ' + str(avg_depth))

### Average the points given to all
avg_all = round(df['points'].mean(),2)
print('Average points for all wines: ' + str(avg_all))

print('Average point difference: ' + str(round(abs(avg_depth - avg_all),2)))

Number of wines with "depth": 2560
Average points having "depth": 90.11
Average points for all wines: 88.45
Average point difference: 1.66


### **6. Wines described using the word "fruity".**
### Wines described using the word "<u>fruity</u>" get <u>LESS</u> points on average.
* Reviews with the word "fruity" averaged **87.60** points.
* All wines together averaged 88.45 points.
* Being described with "fruity" resulted in an average 0.85 point disadvantage.
* The number of wines that have the word "fruity" in their descriptions is 9455.

In [ ]:
### If the description contains the word "fruity", mark True in a new column named 'described_fruity'
df['described_fruity'] = df['description'].str.contains(r'[Ff]ruity')

### Count the number of rows with "fruity"
print('Number of wines with "fruity": ' + str(df[df['described_fruity'] == True].shape[0]))

### Average the points given to those with "fruity".
avg_fruity = round(df[df['described_fruity'] == True]['points'].mean(),2)
print('Average points having "fruity": ' + str(avg_fruity))

### Average the points given to all   # Redunant now, 'cuz avg_all is already set
avg_all = round(df['points'].mean(),2)
print('Average points for all wines: ' + str(avg_all))

print('Average point difference: ' + str(round(abs(avg_fruity - avg_all),2)))

Number of wines with "fruity": 9455
Average points having "fruity": 87.6
Average points for all wines: 88.45
Average point difference: 0.85


### **7. Wines described using the word "herbal".**
### Wines described using the word "<u>herbal</u>" get <u>LESS</u> points on average.**
* Reviews with the word "herbal" averaged **87.44** points.
* All wines together averaged 88.45 points.
* Being described with "herbal" resulted in an average 1.01 point disadvantage.
* The number of wines that have the word "herbal" in their descriptions is 5620.:

In [ ]:
### If the description contains the word "herbal", mark True in a new column named 'described_herbal'
df['described_herbal'] = df['description'].str.contains(r'[Hh]erbal')

### Count the number of rows with "herbal"
print('Number of wines with "herbal": ' + str(df[df['described_herbal'] == True].shape[0]))

### Average the points given to those with "herbal".
avg_herbal = round(df[df['described_herbal'] == True]['points'].mean(),2)
print('Average points having "herbal": ' + str(avg_herbal))

### Average the points given to all   # Redunant now, 'cuz avg_all is already set
avg_all = round(df['points'].mean(),2)
print('Average points for all wines: ' + str(avg_all))

print('Average point difference: ' + str(round(abs(avg_herbal - avg_all),2)))

Number of wines with "herbal": 5620
Average points having "herbal": 87.44
Average points for all wines: 88.45
Average point difference: 1.01


### Bonus data:
8 different wines have descriptions that include all three of the words: "depth", "fruity", & "herbal".
* Their average score was 86.88.
* Having all three words in their description resulted in a lower average than any of them on their own.
* This group's highest score was 89.

In [ ]:
### Make dataframe of only wines having "depth", "fruity", and "herbal"
defrhe_wines = df[(df['described_depth'] == True) & (df['described_fruity'] == True) & (df['described_herbal'] == True)]

### Count the number of rows with "depth", "herbal", and "fruity"
print('Number of wines with "depth", "fruity", and "herbal": ' + str(defrhe_wines.shape[0]))

### Average the points given to those with all three
print('Average points of wines having all three: ' + str(round(defrhe_wines['points'].mean(),2)))

### Set max_colwidth to the length of the longest description so that the entire description can be read.
pd.set_option('display.max_colwidth', defrhe_wines['description'].str.len().max())

### Show those wines
print('\nThose wines and their descriptions:\n')
display(defrhe_wines[['title', 'description', 'points', 'price']].sort_values(by = 'points', ascending = False).reset_index(drop = True))

Number of wines with "depth", "fruity", and "herbal": 8
Average points of wines having all three: 86.88

Those wines and their descriptions:



title  \
0      Château Bousquette 2013 Tradition Red (Saint-Chinian)   
1         Kenwood 1997 Nuns Canyon Zinfandel (Sonoma Valley)   
2                     Château l' Oiseau 2016  Bordeaux Blanc   
3       Ken Forrester 2011 Petit Chenin Blanc (Western Cape)   
4                            Four Aces 2009 Shiraz (Mendoza)   
5                      Robert Mondavi 1997 Merlot (Carneros)   
6  Calmel & Joseph 2015 Villa Blanche Chardonnay (Pays d'Oc)   
7  Calmel & Joseph 2015 Villa Blanche Chardonnay (Pays d'Oc)   

                                                                                                                                                                                                                                                                                                                                                             description  \
0                               Made from organically grown grapes, this blend of 30% Syrah, 30% Mourvèdre, 30% Grenache and 10% Carignan boasts a pleasant balance of ripe black-fruit tones and earthy, herbal hints of garrigue. Hints of violet and lightly toasted oak add depth and interest throughout, with a lush, fruity finish of blackberry and boysenberry.   
1                                         Compared to the Mazzoni, this baby has less fruity but pepperier, earthier aromas, and herbal ones, too: licorice and chocolate, although there’s some deeply scented blackberry. It seems crisper and more angular, and more tannic, with a great hidden depth of fruit that needs to be coaxed out through short-term aging.   
2                                                                                                                                                                                  Strongly herbal, this Sauvignon Blanc-dominated wine is fruity and crisp, with a good depth of flavor. Refreshing acidity adds a tight edge to this young wine. Drink from late 2017.   
3                                                                                                                                       A fresh and fruity wine that's ready to drink now. This has attractive aromas and flavors of green apple and lime followed by a clean, mouthwatering finish. There's a soft herbal streak that adds depth to the nose and mouth.   
4                                                                                      Hefty and fruity, with depth to the black fruit nose. The palate is a bit sticky and tannic, but it's ripe and driving, with wild raspberry, black plum, tea and spice flavors. Finishes a hint herbal, but on the plus side the wine isn't overextracted or too jammy and sweet.   
5  This is a dark-colored, dense wine, reeking of aromas fresh from the fermenting tank: juicy blackberry and plum and a strong herbal, earthy streak like mulch and mushrooms. It's well made and clean as a whistle, but currently the profile is austere, although some fruity depth on the finish bodes well for the long haul. Cellar for three years before try...   
6                                                                                                                          Delicate herbal accents add depth to the fruity tones of apple flesh, pear and white nectarine. The palate is surprisingly round and toasty, with rich baked apple and peach pit flavors that fall somewhat flat on the midpalate. Drink now.   
7                                                                                                                          Delicate herbal accents add depth to the fruity tones of apple flesh, pear and white nectarine. The palate is surprisingly round and toasty, with rich baked apple and peach pit flavors that fall somewhat flat on the midpalate. Drink now.   

   points  price  
0      89   17.0  
1      88   20.0  
2      87   16.0  
3      87   12.0  
4      87   13.0  
5      87   35.0  
6      85   13.0  
7      85   13.0

### **8. Correlation between points and length of the description.**
### There is a <u>moderate positive relationship</u> between the number of characters in the description and the points given. Their correlation coefficient is: <u>+0.558</u>.
* Maybe when a wine is particulary good, the taster tends to elaborate upon the strong points?

&nbsp;

The character count of these descriptions varies around an average of 242.6.
* The longest description was 829 characters long.
* The shortest description was only 20 characters long.

In [ ]:
### Count the number of characters in each description and save this number in a new column named 'desc_len'
df['desc_len'] = df['description'].str.len()

### Show the correlation between description length and points
print('\nThese are the correlation values between description length and points:\n')
display(df[['desc_len', 'points']].corr())

### Determine and show the description length's minimum, average, and maximum
print('\n\nHere are some stats regarding description length:\n')
display(df['desc_len'].agg(min_length = 'min', avg_length = 'mean', max_length = 'max'))


These are the correlation values between description length and points:



desc_len   points
desc_len   1.00000  0.55776
points     0.55776  1.00000



Here are some stats regarding description length:



min_length     20.000000
avg_length    242.601065
max_length    829.000000
Name: desc_len, dtype: float64

### Bonus Novelty Data:
* The wine with the shortest description length is titled '*Craggy Range 2007 Kidnappers Vineyard Chardonnay (Hawke's Bay)*'.
* It's description reads, '<u>*Imported by Kobrand.*</u>' I guess that's a quick way to describe a wine! It received 88 points.


In [ ]:
### Show the wine that has the shortest description
print('\nThe wine with the shortest description:\n')
display(df[['desc_len', 'description', 'title', 'points', 'price', 'country', 'year', 'taster_name']].sort_values(by = 'desc_len').reset_index(drop = True).head(1))


The wine with the shortest description:



desc_len           description  \
0        20  Imported by Kobrand.   

                                                            title  points  \
0  Craggy Range 2007 Kidnappers Vineyard Chardonnay (Hawke's Bay)      88   

   price      country  year     taster_name  
0   24.0  New Zealand  2007  Joe Czerwinski

### To answer question # 9, we need to do some extraction and cleaning of data to analyze data based on region.
The following few cells will guide us through that process...

In [ ]:
### We want to be able to see more of the 'title' column in the following dataframes
### Set max_colwidth to the length of the longest title.
pd.set_option('display.max_colwidth', df['desc_len'].max())

The region that these wines originated within is often found in their titles surrounded by parenthesis. Some wines have their regions recorded here, some do not. We'll extract what we can and save it to a new column named 'region'.

In [ ]:
### Region data extraction ###

### Extract what's between the parenthesis in the title and store them in the new column 'region'
df['region'] = df['title'].str.extract(r'\((.+)\)')[0]

### Show some samples
print('\033[4m'+'\nRegion extraction;\nStep 1 samples:'+'\033[0m')
display(df.loc[[0, 1, 2, 95, 117, 853, 925, 1037, 1480, 3597, 129915], ['title', 'region']].fillna(''))


Region extraction;
Step 1 samples:


title  \
0                                                             Nicosia 2013 Vulkà Bianco  (Etna)   
1                                                 Quinta dos Avidagos 2011 Avidagos Red (Douro)   
2                                                 Rainstorm 2013 Pinot Gris (Willamette Valley)   
95                                                                   Henry Fessy 2015  Juliénas   
117                                                                    Rideau 2014 Estate Syrah   
853                    Dr. H. Thanisch (Erben Thanisch) 2014 Thanisch Kabinett Riesling (Mosel)   
925     Terre Rouge 2013 Vin Doux Naturel Muscat Blanc à Petits Grains (Shenandoah Valley (CA))   
1037                             WildAire 2012 Spofford Station Syrah (Walla Walla Valley (OR))   
1480       Domaines Barons de Rothschild (Lafite) 2007 Los Vascos Chardonnay (Colchagua Valley)   
3597           Domaines Barons de Rothschild (Lafite) 2010 Sélection Prestige  (Bordeaux Blanc)   
129915                                     Balboa 2005 Mith Red Wine Red (Columbia Valley (WA))   

                                                        region  
0                                                         Etna  
1                                                        Douro  
2                                            Willamette Valley  
95                                                              
117                                                             
853     Erben Thanisch) 2014 Thanisch Kabinett Riesling (Mosel  
925                                     Shenandoah Valley (CA)  
1037                                   Walla Walla Valley (OR)  
1480      Lafite) 2007 Los Vascos Chardonnay (Colchagua Valley  
3597          Lafite) 2010 Sélection Prestige  (Bordeaux Blanc  
129915                                    Columbia Valley (WA)

#### Some of these wines have two sets of parenthesis in their titles, so the first process was unable to get all of their regions cleanly. We'll extract the region from these partially extracted strings.

 _Note: I know I didn't need to extract any further to get the answer to question #9. All of the Sicily & Sardinia wines' regions extracted just fine using the first extraction step. I went further as a challenge to myself and as an excersize in learning more python._

In [ ]:
### More region data extraction and cleaning ###

# Extract from the second set of parenthesis and replace the region values that meet these conditions
df.loc[
    (df['region'].str.contains(r'\(WA\)') == False)
    & (df['region'].str.contains(r'\(CA\)') == False)
    & (df['region'].str.contains(r'\(OR\)') == False)
    & (df['region'].notna())
    & (df['region'].str.contains(r'\('))
    , 'region'
  ] = df.loc[
    (df['region'].str.contains(r'\(WA\)') == False)     # Regions in (WA) already extracted correctly. Skip them.
    & (df['region'].str.contains(r'\(CA\)') == False)   # Regions in (CA) already extracted correctly. Skip them.
    & (df['region'].str.contains(r'\(OR\)') == False)   # Regions in (OR) already extracted correctly. Skip them.
    & (df['region'].notna())                            # Can't extract anything from NaN
    & (df['region'].str.contains(r'\('))                # There has to be another opening parenthesis to still have a region that needs further extraction
    , 'region'
  ].str.extract(r'\((.+)$')[0]  # Extract what's between the '(' and the end of the line

In [ ]:
### Show the status of the samples
print('\033[4m'+'\nRegion extraction;\nStep 2 samples:'+'\033[0m')
display(df.loc[[0, 1, 2, 95, 117, 853, 925, 1037, 1480, 3597, 129915], ['title', 'region']].fillna(''))


Region extraction;
Step 2 samples:


title  \
0                                                             Nicosia 2013 Vulkà Bianco  (Etna)   
1                                                 Quinta dos Avidagos 2011 Avidagos Red (Douro)   
2                                                 Rainstorm 2013 Pinot Gris (Willamette Valley)   
95                                                                   Henry Fessy 2015  Juliénas   
117                                                                    Rideau 2014 Estate Syrah   
853                    Dr. H. Thanisch (Erben Thanisch) 2014 Thanisch Kabinett Riesling (Mosel)   
925     Terre Rouge 2013 Vin Doux Naturel Muscat Blanc à Petits Grains (Shenandoah Valley (CA))   
1037                             WildAire 2012 Spofford Station Syrah (Walla Walla Valley (OR))   
1480       Domaines Barons de Rothschild (Lafite) 2007 Los Vascos Chardonnay (Colchagua Valley)   
3597           Domaines Barons de Rothschild (Lafite) 2010 Sélection Prestige  (Bordeaux Blanc)   
129915                                     Balboa 2005 Mith Red Wine Red (Columbia Valley (WA))   

                         region  
0                          Etna  
1                         Douro  
2             Willamette Valley  
95                               
117                              
853                       Mosel  
925      Shenandoah Valley (CA)  
1037    Walla Walla Valley (OR)  
1480           Colchagua Valley  
3597             Bordeaux Blanc  
129915     Columbia Valley (WA)

We were unable to get region data on 10,522 (8.1% of 129,971) of these wines. ☹

In [ ]:
### Determine how many records we were unable to get region data from
print('We were unable to get region data on ' + str(df['region'].isna().sum()) + ' of these wines.')
print(str(round(100 * df['region'].isna().sum() / df.shape[0],2)) + '% of ' + str(df.shape[0]) + '.')

We were unable to get region data on 10522 of these wines.
8.1% of 129971.


### **9. The best region for wine in the province of Sicily & Sardinia.**
### <u>Faro</u> is the region in Sicily & Sardinia, Italy that produces the best wine, on average.
Again, there is a caveat with this one, too. <u>There was only one wine reviewed that came from Faro.</u> It scored 94 points. Having a sample size of one, we can't really say that Faro is the region in Sicily & Sardinia that produces the highest scoring wine, can we?

**Passito di Pantelleria** has the highest average score of wines from Sicily & Sardinia that have 10 or more wines reviewed. This region scored an average of	91.36 points.

In [ ]:
### Create a dataframe of average points grouped by 'region' & 'province', include the count of wines in each
rpp = df[['region', 'points', 'province', 'country']].groupby(['region', 'province', 'country']).agg(avg_points = ('points', 'mean'), wines_count = ('region', 'count')).reset_index()

### Show the top 10 regions by average points limited to the 'province' of 'Sicily & Sardina'
print('\nThe top 10 regions by average points in Sicily & Sardina are:\n')
display(rpp[rpp['province'] == 'Sicily & Sardinia'].sort_values(by = 'avg_points', ascending = False).reset_index(drop = True).head(10))


The top 10 regions by average points in Sicily & Sardina are:



region           province country  avg_points  wines_count
0                    Faro  Sicily & Sardinia   Italy   94.000000            1
1         Moscato di Noto  Sicily & Sardinia   Italy   92.000000            2
2      Contessa Entellina  Sicily & Sardinia   Italy   91.857143            7
3                 Alghero  Sicily & Sardinia   Italy   91.500000            4
4  Passito di Pantelleria  Sicily & Sardinia   Italy   91.363636           11
5       Nasco di Cagliari  Sicily & Sardinia   Italy   91.000000            1
6      Contea di Sclafani  Sicily & Sardinia   Italy   90.333333            6
7   Malvasia delle Lipari  Sicily & Sardinia   Italy   90.250000            4
8                   Eloro  Sicily & Sardinia   Italy   90.200000            5
9  Moscato di Pantelleria  Sicily & Sardinia   Italy   90.000000            1

### Bonus Data:

* Of the regions from which 10 or more wines were reviewed, **Corton** in the province of **Burgundy, France** had the highest average score, 93.32 points.
* Corton had 28 regional wines reviewed in this set.
* **France** is where 7 of the top 10 regions for average points on wine having at least 10 wines reviewed are located.

In [ ]:
print('\nThe top 10 regions by average points having at least 10 reviews:\n')
display(rpp[(rpp['wines_count'] >= 10)].sort_values(by = 'avg_points', ascending = False).reset_index(drop = True).head(10))


The top 10 regions by average points having at least 10 reviews:



region      province   country  avg_points  wines_count
0              Corton      Burgundy    France   93.321429           28
1             Madeira       Madeira  Portugal   93.230769           13
2     Clos de Vougeot      Burgundy    France   93.100000           10
3  Puligny-Montrachet      Burgundy    France   92.920000           75
4   Fort Ross-Seaview    California        US   92.546875           64
5        Saint-Julien      Bordeaux    France   92.484375           64
6              Barsac      Bordeaux    France   92.347826           23
7           Gattinara      Piedmont     Italy   92.304348           23
8            Pauillac      Bordeaux    France   92.228916           83
9         Savennières  Loire Valley    France   92.176471           17

### **Conclusion:**
A large amount of data is recorded on wine. Wine has a rich history, seemingly endless varieties, worldwide crafters, and a passionate community of enthusiasts. Great wines have varying qualities and can come from many different regions.


To pick a truly great wine, consider those from England, India, Austria, or certain regions in France.


Look for wines with 'depth', be wary of wines that are 'herbal', and be a little cautious of 'fruity' wines. Avoid wines with 'depth' that are both 'herbal' and 'fruity' - We're not looking for a trifecta on these aspects.


The more expensive wines are not always better rated wines, so don't jump to that conclusion.


If the wine taster gave a lengthy description, that may be an indicator of quality. And, if you find that Alexander Peartree wrote the review, consider padding its score with another point or two, he may be a little stingy with the points he awards.


2013 was a good vintage year for wines on average, but many may have 'expired' by now.


Of course, you can do what most consumers do - Just grab the bottle in your price range with the coolest looking label!
